nifi-nar-bundles/nifi-hbase-bundle/nifi-hbase-processors/src/main/java/org/apache/nifi/hbase/PutHBaseRecord.java

```diff
@@ -0,0 +1,323 @@
+                .asControllerService(RecordReaderFactory.class);
+        List<PutFlowFile> flowFiles = new ArrayList<>();
+        final String tableName = context.getProperty(TABLE_NAME).evaluateAttributeExpressions(flowFile).getValue();
+        final String rowFieldName = context.getProperty(ROW_FIELD_NAME).evaluateAttributeExpressions(flowFile).getValue();
+        final String columnFamily = context.getProperty(COLUMN_FAMILY).evaluateAttributeExpressions(flowFile).getValue();
+        final String fieldEncodingStrategy = context.getProperty(FIELD_ENCODING_STRATEGY).getValue();
+        final String complexFieldStrategy = context.getProperty(COMPLEX_FIELD_STRATEGY).getValue();
+        final String rowEncodingStrategy = context.getProperty(ROW_ID_ENCODING_STRATEGY).getValue();
+
+        final long start = System.nanoTime();
+        int index = 0;
+        int columns = 0;
+        boolean failed = false;
+        String startIndexStr = flowFile.getAttribute("restart.index");
+        int startIndex = -1;
+        if (startIndexStr != null) {
+            startIndex = Integer.parseInt(startIndexStr);
+        }
+
+        PutFlowFile last  = null;
+        try (RecordReader reader = recordParserFactory.createRecordReader(flowFile, session.read(flowFile), getLogger())) {
+            Record record;
+            if (startIndex >= 0) {
+                while ( index++ < startIndex && (reader.nextRecord()) != null) {}
+            }
+
+            while ((record = reader.nextRecord()) != null) {
+                PutFlowFile putFlowFile = createPut(context, record, reader.getSchema(), flowFile, rowFieldName, columnFamily, fieldEncodingStrategy, rowEncodingStrategy, complexFieldStrategy);
+                flowFiles.add(putFlowFile);
+                index++;
+
+                if (flowFiles.size() == batchSize) {
+                    columns += addBatch(tableName, flowFiles);
+                    last = flowFiles.get(flowFiles.size() - 1);
+                    flowFiles = new ArrayList<>();
+                }
+            }
+            if (flowFiles.size() > 0) {
+                columns += addBatch(tableName, flowFiles);
+                last = flowFiles.get(flowFiles.size() - 1);
+            }
+        } catch (Exception ex) {
+            getLogger().error("Failed to put records to HBase.", ex);
+            failed = true;
+        }
+
+        if (!failed) {
+            sendProvenance(session, flowFile, columns, System.nanoTime() - start, last);
+            flowFile = session.removeAttribute(flowFile, "restart.index");
+            session.transfer(flowFile, REL_SUCCESS);
+        } else {
+            String restartIndex = Integer.toString(index - flowFiles.size());
+            flowFile = session.putAttribute(flowFile, "restart.index", restartIndex);
+            sendProvenance(session, flowFile, columns, System.nanoTime() - start, last);
```

Comment:
Should we wrap this in a conditional like "if columns > 0" then send provenance? 

This would stop us from reporting a provenance event for cases where we didn't send anything successfully.